src/AbstractMenu.js

```diff
@@ -131,50 +131,77 @@ export default class AbstractMenu extends Component {
             return i === currentIndex ? null : i;
         }
 
-        const currentIndex = children.indexOf(selectedItem);
+        const currentIndex = selectedItem ? selectedItem.index : -1;
         const nextEnabledChildIndex = findNextEnabledChildIndex(currentIndex);
 
         if (nextEnabledChildIndex !== null) {
             this.setState({
-                selectedItem: children[nextEnabledChildIndex],
+                selectedItem: {
+                    index: nextEnabledChildIndex,
+                    // We need to know the type of the selected item, so we can
+                    // check it during render and tryToOpenSubMenu.
+                    type: children[nextEnabledChildIndex].type
+                },
                 forceSubMenuOpen: false
             });
         }
     };
 
-    onChildMouseMove = (child) => {
-        if (this.state.selectedItem !== child) {
-            this.setState({ selectedItem: child, forceSubMenuOpen: false });
+    onChildMouseMove = (child, itemIndex) => {
+        if (this.state.selectedItem === null || this.state.selectedItem.index !== itemIndex) {
+            this.setState({
+                selectedItem: {
+                    index: itemIndex,
+                    type: child.type
+                },
+                forceSubMenuOpen: false
+            });
         }
     };
 
     onChildMouseLeave = () => {
         this.setState({ selectedItem: null, forceSubMenuOpen: false });
     };
 
-    renderChildren = children => React.Children.map(children, (child) => {
-        const props = {};
-        if (!React.isValidElement(child)) return child;
-        if ([MenuItem, this.getSubMenuType()].indexOf(child.type) < 0) {
-            // Maybe the MenuItem or SubMenu is capsuled in a wrapper div or something else
-            props.children = this.renderChildren(child.props.children);
-            return React.cloneElement(child, props);
-        }
-        props.onMouseLeave = this.onChildMouseLeave.bind(this);
-        if (child.type === this.getSubMenuType()) {
-            // special props for SubMenu only
-            props.forceOpen = this.state.forceSubMenuOpen && (this.state.selectedItem === child);
-            props.forceClose = this.handleForceClose;
-            props.parentKeyNavigationHandler = this.handleKeyNavigation;
-        }
-        if (!child.props.divider && this.state.selectedItem === child) {
-            // special props for selected item only
-            props.selected = true;
-            props.ref = (ref) => { this.seletedItemRef = ref; };
+    renderChildren = (children, childIndexRef = { value: -1 }) =>
```

Comment:
We are using `childIndexRef` to be able to construct the index properly because this is a recursive function and we need to pass this as a reference. Otherwise, `// Maybe the MenuItem or SubMenu is capsuled in a wrapper div or something else` case doesn't work. And we need it to be working. Otherwise `<div><MenuItem /></div>` won't work.

Also, this function is a bit overcomplicated probably, but I didn't want to change everything.

Reply:
Can you please write this down to a comment?

Probably it would be easier to have an iterative algorithm rather than recursive for this, but I agree with you that we don't want to change everything :-)